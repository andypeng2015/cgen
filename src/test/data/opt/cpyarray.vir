module foo

export function $foo(l %arr, l %src, w %num) {
@entry:
  %i = copy.w 0_w
  jmp @loop
@loop:
  %c = lt.w %i, %num
  br %c, @body, @end
@body:
  %ptr = zext.l %i
  %ptr = mul.l %ptr, 4_l
  %ptr = add.l %ptr, %src
  %w = ld.w %ptr
  %ptr = zext.l %i
  %ptr = mul.l %ptr, 4_l
  %ptr = add.l %ptr, %arr
  st.w %w, %ptr
  %i = add.w %i, 1_w
  jmp @loop
@end:
  ret
}
