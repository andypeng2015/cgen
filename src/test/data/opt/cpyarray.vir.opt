module foo

export function $foo(l %arr, l %src, w %num) {
@2:
  jmp @3(0x0_w)
@3(%i.2):
  %6 = lt.w %i.2, %num ; @23
  br %6, @5, @6
@5:
  %7 = zext.l %i.2 ; @24
  %8 = lsl.l %7, 0x2_l ; @25
  %9 = add.l %8, %src ; @26
  %w.1 = ld.w %9 ; @11
  %10 = add.l %8, %arr ; @27
  st.w %w.1, %10 ; @15
  %11 = add.w %i.2, 0x1_w ; @28
  jmp @3(%11)
@6:
  ret
}
