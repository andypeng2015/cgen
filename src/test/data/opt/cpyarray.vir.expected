module foo

export function $foo(l %arr, l %src, w %num) {
@2:
  jmp @3(0x0_w)
@3(%i.2):
  %0 = lt.w %i.2, %num ; @17
  br %0, @5, @6
@5:
  %1 = zext.l %i.2 ; @18
  %2 = lsl.l %1, 0x2_l ; @19
  %3 = add.l %2, %src ; @20
  %w.1 = ld.w %3 ; @11
  %4 = add.l %2, %arr ; @21
  st.w %w.1, %4 ; @15
  %5 = add.w %i.2, 0x1_w ; @22
  jmp @3(%5)
@6:
  ret
}
